icu4j/main/core/src/main/java/com/ibm/icu/text/DateTimePatternGenerator.java

```diff
@@ -173,15 +167,31 @@ private void initData(ULocale uLocale, boolean skipStdPatterns) {
 
     private void addICUPatterns(PatternInfo returnInfo, ULocale uLocale) {
         // first load with the ICU patterns
-        for (int i = DateFormat.FULL; i <= DateFormat.SHORT; ++i) {
-            SimpleDateFormat df = (SimpleDateFormat) DateFormat.getDateInstance(i, uLocale);
-            addPattern(df.toPattern(), false, returnInfo);
-            df = (SimpleDateFormat) DateFormat.getTimeInstance(i, uLocale);
-            addPattern(df.toPattern(), false, returnInfo);
-
-            if (i == DateFormat.SHORT) {
-                consumeShortTimePattern(df.toPattern(), returnInfo);
+        ICUResourceBundle rb = (ICUResourceBundle) UResourceBundle.getBundleInstance(ICUData.ICU_BASE_NAME, uLocale);
+        String calendarTypeToUse = getCalendarTypeToUse(uLocale);
+        // TODO: this is a HACK to work around ICU-22668/CLDR-17892!  Remove once they're fixed!
+        if (calendarTypeToUse.equals("iso8601")) {
+            calendarTypeToUse = "gregorian";
+        }
```

Comment:
ICU-22668 applies to C/C++. Why is this comment and hack only in Java?

Reply:
I added that because I had a failing unit test on the Java side, but not on the C++ side.  The failing unit test is `DateTimeGeneratorTest.testIso8601()`, which was added by @pedberg-icu and purports to address ICU-21839, which sounds like the same thing I reported in ICU-22668/CLDR-17892.  Peter claims the problem only ever happened in Java, and that it went away by itself (he thinks because of my sideways-inheritance changes), and so he'd only added this unit test to make sure it didn't come back.

So what I'm getting from this is that there was something about calling back into `SimpleDateFormat` to get the "standard" patterns that was sidestepping the ISO 8601 problem, and that my getting the patterns straight out of the resource bundle brought it back.  Maybe there's a subtle difference in that logic between C++ and Java that meant we didn't ever see the symptom in C++.  I haven't tried it, but I'm going to guess my change causes the problem to manifest in C++ as well.

So...

I can port that unit test to C++ and (if necessary) add the "hack" there too.

Or...

I can remove the test from the Java side and assume this is just going to be broken until we address ICU-22668/CLDR-17892.

I prefer option 2.  I'm guessing you're going to want option 1.

Reply:
Messy... I can live with either option.

Reply:
Well, yuck.  I ported that test over to ICU4C, and it passes.  So we've got a subtle difference between C and J.  I added a comment on the J side explaining this, but am not filing any more tickets because I'm hoping the problem goes away once and for all when CLDR-17892 gets fixed.